### ISSUE TITLE
Confusing `struct/c` error complains about a missing accessor instead of showing me the actual problem

### ISSUE THREAD

**Issue by zyrolasting** [racket/racket#3892]
**What version of Racket are you using?**
8.0 [cs]

**What program did you run?**

```
(module _ racket (struct/c S any/c) (struct S (a))
```

**What should have happened?**
An error complaining that the corresponding `struct` should appear before `struct/c`.

**If you got an error message, please include it here.**
`S-a: undefined`

**Please include any other relevant details**
I expect `struct/c` to complain about appearing before `struct`, without mentioning any accessor names.

As for why this is confusing: If I use apply accessors with long names, my first impression from the current error is that I misspelled one of them. Once I see that I spelled everything correctly, I'm wondering where this error could possibly come from. Nothing points me to the fact I put `struct/c` in the wrong place.

**Comment by rfindler** [comment]
I wanted to confirm what you're asking for. When we write something like this:

```
#lang racket
(define a (+ b 5))
(define b 1)
```

we don't get a message that complains that `b` should be defined before `a`, right? In the current setup, the error you mention above is boiling down to an error like the program in this comment.

It might be possible to improve the error message to at least talk about something that's in the program you wrote (instead of something it expanded into) but I don't think we're going to be able to offer the suggestion to reorder the definitions without a more sweeping change.

**Comment by rfindler** [comment]
With the commit above, the error message for this program:

```
#lang racket
(struct/c s integer?)
(struct s (x))
```

changes from:

```
s-x: undefined;
 cannot reference an identifier before its definition
  in module: "/Users/robby/tmp.rkt"
```

to

```
struct s: undefined;
 cannot reference struct in struct/c before its definition
```

This is an improvement in some ways but I see that we also lose the helpful "in module" line. I'm not sure how to get that back without regexp hackery on the exn message string. Maybe that's a reasonable thing to do in this case, tho?

**Comment by zyrolasting** [comment]
Couldn't you preserve the continuation marks from the original exception instead of creating new ones on line 778?

**Comment by rfindler** [comment]
Yes, preserving them would be better. It does not affect the in-module line, no.

**Comment by rfindler** [comment]
I'm not confident in this change because catching `exn:fail:contract:variable` exceptions to redirect the error message seems suboptimal.

**Comment by zyrolasting** [comment]
I understand. Is there a way to check the transformer bindings in advance and check if they would be defined at phase 0 in advance? If not, the only idea I have is to expand to an existential check on the constructor and interpret the exception it raises in a special way.

**Comment by rfindler** [comment]
Doing an analysis to see if something is a forward reference is possible but doesn't seem like the right tool for this job.

**Comment by rfindler** [comment]
Just in general the idea of catching an error and changing the message doesn't seem great. It would be better to have some direct way to ask "is this wrong?" and then go from there, rather than catching and rewriting the message.

**Comment by mflatt** [comment]
I don't think there's a requirement that the names constructed by `struct/dc` are defined by `struct`. So, as a perhaps extreme example, one of the names could be defined as a macro that expands to some expression that references some `letrec`-bound variable too early. Is it a problem if that gets rewritten to a claim that the struct is not yet defined?

**Comment by rfindler** [comment]
Yes, that's a good point! Much better than mine. If there were some way to tell that `struct/c`'s first argument was connected to a struct definition then we could specialize the error in that case, but I'm not sure that information is available.

**Comment by zyrolasting** [comment]
> @zyrolasting: If not, the only idea I have is to expand to an existential check on the constructor and interpret the exception it raises in a special way.
> @mflatt: one of the names could be defined as a macro that expands to some expression that references some letrec-bound variable too early.
> @rfindler: (to @mflatt) Yes, that's a good point!

Flashbacks to https://www.youtube.com/watch?v=k1tsGGz-Qw0 aside, I think we're in agreement here. :laughing:
